### PR TITLE
[Threading] Fix threading package defaults (take 2)

### PR DESCRIPTION
**Pull request by al45tair**
These are better done via the SwiftConfigureSDK mechanism rather than how I was doing them previously.  Additionally, I've changed the way that the swift-threading-package option works.  In addition to specifying just a single package name, you can specify it as a CMake list (i.e. separated by semicolons) of colon-separated `sdk:package` pairs, e.g. `osx:darwin;linux:pthreads`.  You can also override it for all SDKs and then specify for a given SDK; specifications for a particular SDK take precedence over the global override.  For instance `pthreads;osx:darwin` says to use `pthreads` except on the OS X SDK where we should use `darwin`.

### CODE REVIEW

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
preset=stdlib_S_standalone_minimal_macho_x86_64,build,test
@swift-ci please test with toolchain and preset

File: utils/build-script-impl [hunk 1]
```diff
@@ -2227,8 +2227,8 @@ for host in "${ALL_HOSTS[@]}"; do
                     )
                 fi
 
-                if [[ "${SWIFT_THREADING_PACKAGE}" ]] ; then
-                    case "${SWIFT_THREADING_PACKAGE}" in
+                if [[ "${SWIFT_STDLIB_THREADING_PACKAGE}" ]] ; then
```
Comment (buttaface) [inline utils/build-script-impl:2230]: If this is an empty string, I believe bash evaluates this as false and skips this block, so you'll need to correct the list of  cases below.

**Review by finagolfin — CHANGES_REQUESTED** [review]
Overall, looks good, but a few nits.

**Comment by al45tair** [comment]
> Overall, looks good, but a few nits.

:-) I wasn't quite ready for it to be reviewed yet, but thanks for the comments — they were a helpful reminder that I needed to revisit a few things first.

File: cmake/modules/AddSwiftUnittests.cmake [hunk 2]
```diff
@@ -54,7 +53,11 @@ function(add_swift_unittest test_dirname)
     target_compile_definitions("${test_dirname}" PRIVATE
       SWIFT_STDLIB_SINGLE_THREADED_CONCURRENCY)
   endif()
-  threading_package_name("${SWIFT_HOST_VARIANT_SDK}" _threading_package)
+  if(SWIFT_STDLIB_THREADING_PACKAGE)
+    string(TOUPPER "${SWIFT_STDLIB_THREADING_PACKAGE}" _threading_package)
```
[the remark this answers is not in the document]
Comment (al45tair) [reply cmake/modules/AddSwiftUnittests.cmake]: There's a complication with the new mechanism in that setting it for an SDK affects both toolchain and stdlib compilations if they happen to be using the same SDK. This was the first thing that came to mind to work around that, but of course it means the name is wrong. But I have something much better now.

File: stdlib/cmake/modules/AddSwiftStdlib.cmake [hunk 3]
```diff
@@ -355,7 +354,11 @@ function(_add_target_variant_c_compile_flags)
     list(APPEND result "-DSWIFT_STDLIB_SINGLE_THREADED_CONCURRENCY")
   endif()
 
-  threading_package_name("${CFLAGS_SDK}" _threading_package)
+  if(SWIFT_STDLIB_THREADING_PACKAGE)
+    string(TOUPPER "${SWIFT_STDLIB_THREADING_PACKAGE}" _threading_package)
```
[the remark this answers is not in the document]
Comment (al45tair) [reply stdlib/cmake/modules/AddSwiftStdlib.cmake]: These are gone anyway. I've changed the way the override works.

File: utils/build-script-impl (hunk 1 above)
Comment (al45tair) [reply utils/build-script-impl]: I don't believe this is relevant any more either.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by buttaface** [comment]
Nice, thanks for adding the list of overrides, that's useful for cross-compilation.

File: utils/build-script-impl [hunk 4]
```diff
@@ -206,7 +206,7 @@ KNOWN_SETTINGS=(
     swift-stdlib-has-dladdr                       "1"               "whether to build stdlib assuming the runtime environment provides dladdr API"
     swift-stdlib-supports-backtrace-reporting     ""                "whether to build stdlib assuming the runtime environment provides the backtrace(3) API, if not set defaults to true on all platforms except for Cygwin, Haiku and wasm"
     swift-runtime-static-image-inspection         "0"               "whether to build stdlib assuming the runtime environment only supports a single runtime image with Swift code"
-    swift-threading-package                       ""                "which threading package to use for Swift; valid values are empty string (default based on platform), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
+    swift-threading-package                  ""                     "override the threading package for the host build; this is either a single package or a semicolon-separated list of sdk:package pairs.  Valid packages are empty string (no override), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
```
Comment (buttaface) [inline utils/build-script-impl:209]: Nit, move the quotes so it lines up with the other values in one column.

File: utils/build-script-impl [hunk 5]
```diff
@@ -206,7 +206,7 @@ KNOWN_SETTINGS=(
     swift-stdlib-has-dladdr                       "1"               "whether to build stdlib assuming the runtime environment provides dladdr API"
     swift-stdlib-supports-backtrace-reporting     ""                "whether to build stdlib assuming the runtime environment provides the backtrace(3) API, if not set defaults to true on all platforms except for Cygwin, Haiku and wasm"
     swift-runtime-static-image-inspection         "0"               "whether to build stdlib assuming the runtime environment only supports a single runtime image with Swift code"
-    swift-threading-package                       ""                "which threading package to use for Swift; valid values are empty string (default based on platform), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
+    swift-threading-package                       ""                     "override the threading package for the host build; this is either a single package or a semicolon-separated list of sdk:package pairs.  Valid packages are empty string (no override), 'pthreads', 'darwin', 'linux', 'win32', 'c11', 'none'"
```
Comment (buttaface) [inline utils/build-script-impl:209]: Heh, now the last column with the description is misaligned. 😉

Comment (al45tair) [reply utils/build-script-impl]: 😂 Fixed now.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by buttaface** [comment]
@compnerd, would you review?

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 6]
```diff
@@ -101,6 +102,34 @@ function(remove_sdk_unsupported_archs name os sdk_path architectures_var)
...
+# Work out which threading package to use by consulting SWIFT_THREADING_PACKAGE
+function(find_threading_package sdk default package_var)
+  set(global_override)
+  foreach(elt ${SWIFT_THREADING_PACKAGE})
+    string(REPLACE ":" ";" elt_list "${elt}")
+    list(LENGTH elt_list elt_list_len)
+    if(elt_list_len EQUAL 1)
+      list(GET elt_list 0 global_override)
+      string(TOLOWER "${global_override}" global_override)
+    else()
```
Comment (compnerd) [inline cmake/modules/SwiftConfigureSDK.cmake:114]: Should we have a case to ensure that someone doesn't form an invalid input with multiple `:`?

File: cmake/modules/SwiftConfigureSDK.cmake [hunk 7]
```diff
@@ -101,6 +102,34 @@ function(remove_sdk_unsupported_archs name os sdk_path architectures_var)
   set("${architectures_var}" ${architectures} PARENT_SCOPE)
 endfunction()
 
+# Work out which threading package to use by consulting SWIFT_THREADING_PACKAGE
+function(find_threading_package sdk default package_var)
```
Comment (compnerd) [inline cmake/modules/SwiftConfigureSDK.cmake:106]: `get_threading_package` seems like a better name as it is meant to retrieve the value.

File: cmake/modules/SwiftConfigureSDK.cmake (hunk 6 above)
Comment (al45tair) [reply cmake/modules/SwiftConfigureSDK.cmake:114]: Good idea. Done.

**Comment by al45tair** [comment]
@swift-ci Please test

**Comment by al45tair** [comment]
@swift-ci Please test macOS platform

**Review by compnerd — APPROVED** [review]
Seems reasonable to me.  I didn't carefully review the bash bits, and I suppose it is fine as is, though I tend to prefer to write `sdk=${elt%%:*}` and `package=${elt##*:}` for longest prefix and suffix replacement rather than the regular substitution.

**Pull request merged**